### PR TITLE
fix completed courses bug

### PR DESCRIPTION
**Pull request by justineluo**
(if gif is not working, please ctrl+click to open in new tab)
Graduate PR #1 gif

# Description

Fixes a bug where transfer courses added during onboarding weren't correctly being added to the schedule as transfer courses. More specifically, fixes:

1. slowdown due to consecutive awaits for fetching courses (fix is to use Promise.all)
2. incorrect course equality checks
3. double-equals equality checks

Additionally, this cleans up some missing-key warnings, deduplicates some code,
and also adds text for the steps displayed during onboarding.

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe how you tested this PR (both manually and with tests)

Tested locally by selecting completed courses in the onboarding workflow and checking that they show up in transfer credit holder.

I have not written any tests for my code.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code where needed 
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# Outstanding Questions

1. Should we change the strategy used to add completed courses to the schedule? (topological sort)
2. Is the text for steps displayed in the onboarding flow okay to add?

### High-level summary of changes:

- refactor TransferCoursesScreen to be functional component
- pull out shared utils from transfer screen and completed courses screen to course-helpers
- add explicitly controlled SelectableCourse component to GenericOnboarding

### CODE REVIEW

File: frontend/src/Onboarding/CompletedCoursesScreen.tsx [hunk 1]
```diff
@@ -2,48 +2,20 @@ import React, { Component } from "react";
...
-    req.type === "AND" &&
-    req.courses.filter(c => c.type === "COURSE").length
-  ) {
-    return [req.courses as IRequiredCourse[]];
-  } else if (req.type === "AND" || req.type === "OR") {
-    return flatten(req.courses);
-  } else {
-    return [];
-  }
-}
```
Comment (bugsalexander) [inline frontend/src/Onboarding/CompletedCoursesScreen.tsx:46]: these were duplicated in the transfer courses screen, so i moved them to `course-helpers.ts`

File: frontend/src/Onboarding/GenericOnboarding.tsx [hunk 2]
```diff
@@ -41,7 +41,6 @@ const Box = styled.div`
 
 const DotWrapper = styled.div`
   display: flex;
-  width: 256;
```
Comment (bugsalexander) [inline frontend/src/Onboarding/GenericOnboarding.tsx:44]: removed static width on this so that the steps center properly

File: frontend/src/Onboarding/GenericOnboarding.tsx [hunk 3]
```diff
@@ -113,7 +115,12 @@ const theme = createMuiTheme({
   },
 });
 
-const steps = ["", "", "", ""];
+const steps = [
+  "Student Profile",
+  "Completed Courses",
+  "Transfer Courses",
+  "AP/IB Courses",
+];
```
Comment (bugsalexander) [inline frontend/src/Onboarding/GenericOnboarding.tsx:123]: added steps so that users can view what steps they have completed/remaining. we can totally change the text here. changed this due to a warning, see below

File: frontend/src/Onboarding/GenericOnboarding.tsx [hunk 4]
```diff
@@ -151,6 +158,25 @@ const SelectableCourse: React.FC<SelectableCourseProps> = ({
   );
 };
 
+const ControlledSelectableCourse: React.FC<SelectableCourseProps & {
```
Comment (bugsalexander) [inline frontend/src/Onboarding/GenericOnboarding.tsx:161]: The above `SelectableCourse` tries to be an [uncontrolled component](https://reactjs.org/docs/uncontrolled-components.html), but according to the MUI docs, [controlled components take an `onChange`](https://mui.com/components/checkboxes/#controlled).

I had tried using `SelectableCourse` as it had been used before, but for whatever reason I couldn't get it to be uncontrolled (clicking checkbox would not check the component). I'm guessing it's because we pass it an `onChange`? Though I'm not sure, since that seems to be elsewhere... it may also be context dependent (i'm using it in a function component, the others are classes).

TLDR was having trouble getting uncontrolled to work, so I made a controlled variant of `SelectableCourse`. 

If reviewers think it's cleaner, I can also add a `controlled` prop to the existing `SelectableCourse` component so that there's less code duplication.

File: frontend/src/Onboarding/TransferCoursesScreen.tsx [hunk 5]
```diff
@@ -35,253 +30,159 @@ const TitleText = styled.div`
...
-    return [req.courses as IRequiredCourse[]];
-  } else if (req.type === "AND" || req.type === "OR") {
-    return flatten(req.courses);
-  } else {
-    return [];
-  }
-}
-
-interface TransferCoursesScreenProps {
+type TransferCoursesScreenProps = {
```
Comment (bugsalexander) [inline frontend/src/Onboarding/TransferCoursesScreen.tsx:33]: Since I was refactoring anyway, I converted this to be a functional component as well, lmk thoughts on this

File: frontend/src/Onboarding/TransferCoursesScreen.tsx [hunk 6]
```diff
@@ -35,253 +30,159 @@ const TitleText = styled.div`
...
+const TransferCoursesScreenComponent: React.FC<TransferCoursesScreenProps> = () => {
+  const major = useSelector((state: AppState) => getUserMajorFromState(state)!);
+  const completedRequirements = useSelector((state: AppState) =>
+    getCompletedRequirementsFromState(state)
+  );
+  const dispatch = useDispatch();
+  const setCompletedCourses = (completedCourses: ScheduleCourse[]) =>
+    dispatch(setCompletedCoursesAction(completedCourses));
+  const setTransferCourses = (transferCourses: ScheduleCourse[]) =>
+    dispatch(setTransferCoursesAction(transferCourses));
```
Comment (bugsalexander) [inline frontend/src/Onboarding/TransferCoursesScreen.tsx:62]: first time using redux 🤪 so lmk if this looks ok or not

File: frontend/src/Onboarding/TransferCoursesScreen.tsx [hunk 7]
```diff
@@ -35,253 +30,159 @@ const TitleText = styled.div`
...
    * Separate transfer and non transfer courses, convert non transfer completed courses
    * into ScheduleCourses, and populate redux store.
    */
-  async onSubmit() {
-    let completedNonTransfer = this.state.completedNonTransfer.filter(
-      req => !this.state.selectedRequirements.includes(req)
+  const onSubmit = async () => {
+    const filteredCompleted = completedNonTransfer.filter(
+      req => !selectedRequirements.some(c => courseEq(c, req))
     );
```
Comment (bugsalexander) [inline frontend/src/Onboarding/TransferCoursesScreen.tsx:120]: this is what I think was the first bug: `includes` uses ref equality (bad). it may have been fine, but it's generally an antipattern to use ref equality.

File: frontend/src/Onboarding/TransferCoursesScreen.tsx [hunk 8]
```diff
@@ -35,253 +30,159 @@ const TitleText = styled.div`
...
-    let completed: ScheduleCourse[] = [];
-
-    for (let course of completedNonTransfer) {
-      const scheduleCourse = await fetchCourse(
-        course.subject,
-        String(course.classId)
-      );
```
Comment (bugsalexander) [inline frontend/src/Onboarding/TransferCoursesScreen.tsx:127]: (semi) bug # 2 `await` inside of for-loop (bad)

File: frontend/src/Onboarding/TransferCoursesScreen.tsx [hunk 9]
```diff
@@ -35,253 +30,159 @@ const TitleText = styled.div`
...
-   * and if it is being checked, it converts it to a ScheduleCourse and adds it to selected courses
-   * @param e
-   * @param course
-   */
-  async onChecked(e: any, course: IRequiredCourse) {
-    const checked = e.target.checked;
-
-    if (checked) {
-      this.setState({
-        selectedRequirements: [...this.state.selectedRequirements, course],
```
Comment (bugsalexander) [inline frontend/src/Onboarding/TransferCoursesScreen.tsx:161]: bug # 3: this was used to set state multiple times for a single checkbox (ie checking CS2510 and CS2511), and so one of them was getting wiped out (only setting the second of the two, CS2511, as checked).

File: frontend/src/Onboarding/TransferCoursesScreen.tsx [hunk 10]
```diff
@@ -35,253 +30,159 @@ const TitleText = styled.div`
...
-      let reqs = this.state.selectedRequirements.filter(r => r !== course);
-      this.setState({
-        selectedRequirements: reqs,
-      });
-    }
-  }
+    const [completed, transfer] = await Promise.all([
+      fetchAndFilterCourses(filteredCompleted),
+      fetchAndFilterCourses(selectedRequirements),
+    ]);
```
Comment (bugsalexander) [inline frontend/src/Onboarding/TransferCoursesScreen.tsx:93]: fix for await bug is to use promise.all

File: frontend/src/Onboarding/TransferCoursesScreen.tsx [hunk 11]
```diff
@@ -35,253 +30,159 @@ const TitleText = styled.div`
...
-  /**
-   * Renders one course/courseset (if it contains labs/recitiations, and seperated)
-   * @param courses - Course pairings to be rendered
-   */
-  renderCourse(courses: IRequiredCourse[]) {
-    let allCourse = courses.map(course => course.subject + course.classId);
-    return (
-      <SelectableCourse
-        courseText={allCourse.join(" and ")}
-        onChange={e => courses.forEach(course => this.onChecked(e, course))}
```
Comment (bugsalexander) [inline frontend/src/Onboarding/TransferCoursesScreen.tsx:231]: bug # 3 part 2: really this should be _one_ state update for all the courses in the set, not several updates (one per course). the later setState calls will override the beginning ones

File: frontend/src/Onboarding/TransferCoursesScreen.tsx [hunk 12]
```diff
@@ -35,253 +30,159 @@ const TitleText = styled.div`
...
-      )
-      .map((r: IRequiredCourse[]) => this.renderCourse(r));
-  }
+  const CourseSet: React.FC<{ courses: IRequiredCourse[] }> = ({ courses }) => {
+    const filtered = selectedRequirements.filter(
+      r => !courses.some(c => courseEq(r, c))
+    );
+    return (
+      <ControlledSelectableCourse
+        checked={filtered.length !== selectedRequirements.length}
```
Comment (bugsalexander) [inline frontend/src/Onboarding/TransferCoursesScreen.tsx:108]: pass in `checked`, so this is a controlled component

File: frontend/src/utils/completed-courses-helpers.ts [hunk 13]
```diff
@@ -1,32 +1,11 @@
...
   ScheduleYear,
-  SeasonWord,
 } from "../../../common/types";
 import { convertToDNDSchedule } from "./schedule-helpers";
 
-/**
- * Returns the sum of all credits in the courses
- * @param courses
- */
-export function sumCreditsFromList(courses: DNDScheduleCourse[]): number {
```
Comment (bugsalexander) [inline frontend/src/utils/completed-courses-helpers.ts:18]: this was unused, so i deleted it. lmk if i should add it back

File: frontend/src/utils/completed-courses-helpers.ts [hunk 14]
```diff
@@ -84,83 +69,62 @@ export function getNextTerm(is_summer: boolean, classes: DNDScheduleCourse[]) {
   return classes.splice(0, counter);
 }
 
-export function getSimplifiedCourseData(
```
Comment (bugsalexander) [inline frontend/src/utils/completed-courses-helpers.ts:87]: unused pt2

File: frontend/src/utils/completed-courses-helpers.ts [hunk 15]
```diff
@@ -84,83 +69,62 @@ export function getNextTerm(is_summer: boolean, classes: DNDScheduleCourse[]) {
...
-  const termMap: { [key: string]: SeasonWord } = {
-    "10": "fall",
-    "30": "spring",
-    "40": "summer1",
-    "60": "summer2",
+
+  addBlankScheduleYear(years, yearMap, 2022);
+  yearMap[years[0]].fall.classes = [...completedCourses];
+  const [schedule, counter] = convertToDNDSchedule({ years, yearMap }, 0);
+  return { schedule, counter };
```
Comment (bugsalexander) [inline frontend/src/utils/completed-courses-helpers.ts:86]: Question: for this, we can actually do better. my idea is to use topological sort (implemented in json_parser) to create an ordering of courses in prereq-order. thoughts? otherwise for now, we just throw all the completed courses in fall and let the user handle moving them around.

File: scrapers/src/json_converter.ts [hunk 16]
```diff
@@ -34,6 +34,8 @@ export const parseRequirement = (
       return parseRequiredRange(req);
     case "COURSE":
       return parseRequiredSimple(req);
+    case "CREDITS":
+      throw new Error("unimplemented!");
```
Comment (bugsalexander) [inline scrapers/src/json_converter.ts:38]: just to make `yarn tsc` work

**Comment by bugsalexander** [comment]
[Debug why transfer courses from onboarding are partially displayed in transfer holder](https://trello.com/c/18KZppk6/382-debug-why-transfer-courses-from-onboarding-are-partially-displayed-in-transfer-holder)

File: frontend/src/Onboarding/GenericOnboarding.tsx (hunk 4 above)
Comment (aryanshah701) [reply frontend/src/Onboarding/GenericOnboarding.tsx:161]: interesting, why do you think we are doing uncontrolled select components for some areas then(completed course screen for example)? if react suggests controlled components what makes our case unique in requiring an uncontrolled select component?

Comment (aryanshah701) [reply frontend/src/Onboarding/GenericOnboarding.tsx:161]: What do you think about abstracting the component into a general select with the `isControlled` prop but also exporting `UncontrolledSelectableCourse` and `ControlledSelectableCourse` that essentially just wrap around the generic select and pass down true or false for the `isControlled` prop. 

If you find that clean that would be awesome! Otherwise just adding the prop is fine but lets maybe abstract out the common MUI styling and stuff :)

File: frontend/src/Onboarding/TransferCoursesScreen.tsx (hunk 5 above)
Comment (aryanshah701) [reply frontend/src/Onboarding/TransferCoursesScreen.tsx:33]: awesome!!!

File: frontend/src/Onboarding/TransferCoursesScreen.tsx (hunk 7 above)
Comment (aryanshah701) [reply frontend/src/Onboarding/TransferCoursesScreen.tsx:120]: good catch!!

File: frontend/src/Onboarding/TransferCoursesScreen.tsx (hunk 6 above)
Comment (aryanshah701) [reply frontend/src/Onboarding/TransferCoursesScreen.tsx:62]: ooh I like `useDispatch` and `useSelector`! Curious are there any benefits to using these as opposed to `connect` or is this just the more functional/hooks way of doing things?

File: frontend/src/Onboarding/TransferCoursesScreen.tsx (hunk 11 above)
Comment (aryanshah701) [reply frontend/src/Onboarding/TransferCoursesScreen.tsx:231]: sweet!

File: frontend/src/utils/completed-courses-helpers.ts (hunk 15 above)
Comment (aryanshah701) [reply frontend/src/utils/completed-courses-helpers.ts:86]: i think that's an awesome idea, it doesn't seem too bad to implement too! bring this up during standup and @justineluo can add it to our backlog and prioritize it!

File: frontend/src/utils/completed-courses-helpers.ts [hunk 17]
```diff
@@ -84,83 +69,62 @@ export function getNextTerm(is_summer: boolean, classes: DNDScheduleCourse[]) {
...
+): { schedule: DNDSchedule; counter: number } {
   const years: number[] = [];
   const yearMap: { [key: number]: ScheduleYear } = {};
-  const termMap: { [key: string]: SeasonWord } = {
-    "10": "fall",
-    "30": "spring",
-    "40": "summer1",
-    "60": "summer2",
+
+  addBlankScheduleYear(years, yearMap, 2022);
```
Comment (aryanshah701) [inline frontend/src/utils/completed-courses-helpers.ts:83]: question: why do we have only add a 2022 blank schedule year? And why specifically 2022? Do we just dump completed courses into a 2022 year or am i misunderstanding this?

File: scrapers/src/json_converter.ts (hunk 16 above)
Comment (aryanshah701) [reply scrapers/src/json_converter.ts:38]: woah, did yarn tsc not work before or is this related to some change made in this pr??

File: frontend/src/utils/course-helpers.ts [hunk 18]
```diff
@@ -55,3 +57,32 @@ export async function getScheduleCourseCoreqs(
...
+    return flatten(req.courses);
+  } else {
+    return [];
+  }
+}
+
+export const courseToString = (c: {
+  subject: string;
+  classId: number | string;
+}) => `${c.subject}${c.classId}`;
```
Comment (aryanshah701) [inline frontend/src/utils/course-helpers.ts:88]: Thank you!!! I think we have course to string being done in so many places but nobody bothered to standardize it. If you have time and saw some of it would really appreciate it if you could add this function but don't spend more than a few minutes!(hint: line 114 in CompletedCoursesScreen)

**Review by aryanshah701 — COMMENTED** [review]
Amazing stuff!! Loved the adhoc changes to clean things up to whichever files you touched <3. Just have a couple of questions regarding the changes + 1 small refactor request of the Selectable component, otherwise all good to merge!

File: frontend/src/Onboarding/LoginScreen.tsx [hunk 19]
```diff
@@ -58,11 +58,6 @@ const ButtonContainer = styled.div`
 `;
 
 interface ReduxStoreLoginScreenProps {
-  setAcademicYear: (academicYear: number) => void;
-  setGraduationYear: (graduationYear: number) => void;
-  setUserCoopCycle: (coopCycle: string) => void;
-  setSchedules: (schedules: NamedSchedule[]) => void;
-  setPlanStr: (planStr: string) => void;
```
[the remark this answers is not in the document]
Comment (bugsalexander) [reply frontend/src/Onboarding/LoginScreen.tsx:65]: shoutout to @justineluo

**Comment by bugsalexander** [comment]
> interesting, why do you think we are doing uncontrolled select components for some areas then(completed course screen for example)? if react suggests controlled components what makes our case unique in requiring an uncontrolled select component?

not sure. my best guess is that at the time, it was easier to do, or that we weren't aware of doing so.

**Comment by bugsalexander** [comment]
> What do you think about abstracting the component into a general select with the isControlled prop but also exporting UncontrolledSelectableCourse and ControlledSelectableCourse that essentially just wrap around the generic select and pass down true or false for the isControlled prop.

yep! i may also just replace all the instances of the uncontrolled variant, if it's easy enough to do.

File: frontend/src/Onboarding/TransferCoursesScreen.tsx (hunk 6 above)
Comment (bugsalexander) [reply frontend/src/Onboarding/TransferCoursesScreen.tsx:62]: yes, this is just the functional way of doing things. and there aren't "benefits" per se, but the API is slightly different/more flexible, see https://react-redux.js.org/api/hooks#useselector

File: frontend/src/Onboarding/GenericOnboarding.tsx (hunk 4 above)
Comment (bugsalexander) [reply frontend/src/Onboarding/GenericOnboarding.tsx:161]: not sure. my best guess is that at the time, it was easier to do, or that we weren't aware of doing so.

and yep! i may also just replace all the instances of the uncontrolled variant, if it's easy enough to do.

File: scrapers/src/json_converter.ts (hunk 16 above)
Comment (bugsalexander) [reply scrapers/src/json_converter.ts:38]: didn't work before

File: frontend/src/utils/completed-courses-helpers.ts (hunk 17 above)
Comment (bugsalexander) [reply frontend/src/utils/completed-courses-helpers.ts:83]: oh yeah i need to fix this, thanks!

File: frontend/src/utils/course-helpers.ts (hunk 18 above)
Comment (bugsalexander) [reply frontend/src/utils/course-helpers.ts:88]: i'll do a look over!

File: frontend/src/Onboarding/GenericOnboarding.tsx (hunk 2 above)
Comment (bugsalexander) [reply frontend/src/Onboarding/GenericOnboarding.tsx:44]: test comment, ignore

File: frontend/src/Onboarding/GenericOnboarding.tsx (hunk 4 above)
Comment (aryanshah701) [reply frontend/src/Onboarding/GenericOnboarding.tsx:161]: sweet thanks!!

File: scrapers/src/json_converter.ts (hunk 16 above)
Comment (aryanshah701) [reply scrapers/src/json_converter.ts:38]: crazy

File: frontend/src/utils/completed-courses-helpers.ts (hunk 17 above)
Comment (bugsalexander) [reply frontend/src/utils/completed-courses-helpers.ts:83]: just tested it out and it seems like currently the years get overwritten to start at graduating year - current years at neu, and dont' actually use this at all. i'll change it to a constant and leave a comment

Comment (bugsalexander) [reply frontend/src/utils/completed-courses-helpers.ts:83]: previously it was getting overwritten in `studentReducer.ts` by `alterScheduleToHaveCorrectYears` but rather than having two tightly-coupled functions, i just pulled out a function. should be good now

File: frontend/src/utils/course-helpers.ts (hunk 18 above)
Comment (bugsalexander) [reply frontend/src/utils/course-helpers.ts:82]: done

**Review by aryanshah701 — APPROVED** [review]
looks good to me!! also thank you so much for the courseToString changes!
